### PR TITLE
[WIP] DialogueChannel can route to a specific host.

### PR DESCRIPTION
**Pull request by jkozlowski**
## Before this PR

Host-specific routing is a concern external to DialogueChannel.

## After this PR
==COMMIT_MSG==
DialogueChannel can route to a specific host.
==COMMIT_MSG==

### CODE REVIEW

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 1]
```diff
@@ -333,14 +311,42 @@ public String toString() {
...
+        }
+    }
+
+    interface InternalDialogueChannel extends Channel {
+        List<Channel> perHostChannels();
+    }
+
+    private static final class InternalDialogueChannelFromDialogueChannel implements InternalDialogueChannel {
+
+        private final DialogueChannel dialogueChannel;
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java:328]: Not sure why we don't just use the DialogueChannel directly. Don't we want one that also implements EndpointChannelFactory?

Comment (jkozlowski) [reply dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java:328]: There might be a better way to do it, but if you have a look at ReloadingClientFactory#getInternalDialogueChannel, we have things like AlwaysThrowingChannel, and the return type must have the same super-type.

Comment (carterkozak) [reply dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java:328]: Got it. Might be helpful to add that detail to InternalDialogueChannel javadoc.

File: dialogue-core/src/test/java/com/palantir/dialogue/core/DialogueChannelTest.java [hunk 2]
```diff
@@ -301,6 +301,9 @@ void nice_tostring() {
                 .isNotEqualTo(channel2.toString());
     }
 
+    @Test
+    public void can_request_executed_on_attachment() {}
```
Comment (carterkozak) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/DialogueChannelTest.java:305]:
```suggestion
    public void can_request_executed_on_attachment() {
        throw new AssertionError("TODO");
    }
```

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 3]
```diff
@@ -316,7 +260,41 @@ public String toString() {
...
+            ServiceConfiguration serviceConf =
+                    ServiceConfigurationFactory.of(block).get(serviceName);
+
+            DialogueChannel dialogueChannel =
+                    cache.getNonReloadingChannel(params, serviceConf, channelName, OptionalInt.empty());
+            return new InternalDialogueChannelFromDialogueChannel(dialogueChannel);
+        });
+    }
+
+    private static final class AlwaysThrowingChannel implements InternalDialogueChannel {
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]:
```suggestion
    private static final class NoUrisDialogueChannel implements InternalDialogueChannel {
```

File: dialogue-core/src/main/java/com/palantir/dialogue/core/HostLimitedChannel.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+interface HostLimitedChannel extends LimitedChannel {
+    HostIdx getHostIdx();
```
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/HostLimitedChannel.java]: Perhaps compose over inherit: `LimitedChannel channel()` and rename to HostAndLimitedChannel or similar?

**Pull request closed without merging**